src/core/mesh/qgstriangularmesh.h

```diff
@@ -118,6 +120,47 @@ class CORE_EXPORT QgsTriangularMesh
 
     QVector<QVector3D> vertexNormals( float vertScale ) const;
 
+    /**
+     * Returns simplified meshes. The first simplified mesh is simplified with a goal of a number of triangles equals to the
+     * number of triangles of the base mesh divised by the reduction factor. For the following mesh the same reduction factor is used with
+     * the prededent goal of number of triangles.
+     * There are as many simplified meshes as necessary to have a the minimum triangles count on the last simplified mesh.
+     *
+     * The caller has to take the ownership of returned meshes.
+     *
+     * \param reductionFactor is the factor used to reduce the number of triangles of the mesh
+     * \param minimumTrianglesCount is the minimal faces count on simplified mesh
+     * \since QGIS 3.14
+     */
+    QVector<QgsTriangularMesh *> simplifyMesh( double reductionFactor, int minimumTrianglesCount = 10 ) const;
+
+    /**
+     * Returns the average size of triangles in map unit. It is calculated using the maximum of the height/width of the
+     * bounding box of each triangles.
+     *
+     * \since QGIS 3.14
+     */
+    double averageTriangleSize() const;
+
+    /**
+     * Returns the corresponding index of level of detail on which this mesh is associated
+     * 0 : base mesh
+     * 1 : first simplified mesh
+     * 2 : second simplified mesh (lower level of detail)
+     * ...
+     * ...
+     *
+     * \since QGIS 3.14
+     */
+    int levelOfDetail() const;
+
+    /**
+     * Returns base triangular used to build this mesh. If this mesh is the base mesh, returns nullptr
+     *
+     * \since QGIS 3.14
+     */
+    const QgsTriangularMesh *baseTriangularMesh() const;
```

Comment:
where is this used? I fear it may return you invalid pointer at some point when you for example update base triangular mesh..?

Reply:
The method is used by the renderer to retrieve the native mesh. Indeed, the part of the native mesh in the extent is retrieved from the triangles in extent. But the simplified mesh doesn't have the same map triangles/natives (for simplified mesh, the map hasn't all the native faces, but only the faces in the centroid of the triangle face, then this map is used for rendering dataset on faces like for base mesh), so we can't used the simplified mesh map to retireved the native mesh in the extent, otherwise, only the native faces which are in the centroid of the simplified mesh will be rendered.
So we need the base triangular mesh to retrieve the native faces in the extent. But this is only used for rendering the native mesh when simplified mesh is used...

And, well, you are right, there is a risk of undefined behaviour if the base mesh is updated when rendering, and it could happen as the renderer is used in another thread... We can't keep this implementation...
But it is quite tricky to fix... severals solutions:
- disabled the native mesh rendering if simplified mesh is rendered. The native mesh will not be rendered except if the zoom is enought to activate the rendering of the base mesh and so, also the native mesh. 
- add a reference to the base triangular mesh in the renderer class. As this will be a shared copy, not a lot of more memory needed. But we will have a reference to the triangular mesh AND to the base triangular mesh. It wil be the same mesh if it is the base mesh which is rendered, with two references. I don't think this solution is very clean, but I don't see other ... The base mesh reference will be used only when the native mesh is rendered.

After writing that, I am more convinced for the first solution that is very simple and not leads to a big regression : if the user activates the simplifification, he may doesn't care a lot about the native mesh, and there is may be not a lot of sens to continue to displaying the native mesh.

Reply:
happy with the changes, can you remove commented out code please?